ghcide/src/Development/IDE/Core/FileStore.hs

```diff
@@ -93,24 +96,39 @@ isFileOfInterestRule = defineEarlyCutoff $ \IsFileOfInterest f -> do
     let res = maybe NotFOI IsFOI $ f `HM.lookup` filesOfInterest
     return (Just $ BS.pack $ show $ hash res, ([], Just res))
 
-getModificationTimeRule :: VFSHandle -> Rules ()
-getModificationTimeRule vfs =
+getModificationTimeRule :: VFSHandle -> (NormalizedFilePath -> Action Bool) -> Rules ()
+getModificationTimeRule vfs isWatched =
     defineEarlyCutoff $ \(GetModificationTime_ missingFileDiags) file -> do
         let file' = fromNormalizedFilePath file
         let wrap time@(l,s) = (Just $ BS.pack $ show time, ([], Just $ ModificationTime l s))
-        alwaysRerun
         mbVirtual <- liftIO $ getVirtualFile vfs $ filePathToUri' file
         case mbVirtual of
-            Just (virtualFileVersion -> ver) ->
+            Just (virtualFileVersion -> ver) -> do
+                alwaysRerun
                 pure (Just $ BS.pack $ show ver, ([], Just $ VFSVersion ver))
-            Nothing -> liftIO $ fmap wrap (getModTime file')
-              `catch` \(e :: IOException) -> do
-                let err | isDoesNotExistError e = "File does not exist: " ++ file'
-                        | otherwise = "IO error while reading " ++ file' ++ ", " ++ displayException e
-                    diag = ideErrorText file (T.pack err)
-                if isDoesNotExistError e && not missingFileDiags
-                    then return (Nothing, ([], Nothing))
-                    else return (Nothing, ([diag], Nothing))
+            Nothing -> do
+                isWF <- isWatched file
+                unless isWF alwaysRerun
```

Comment:
Does this kind of conditional-`alwaysRerun` work? I seem to recall getting very tangled up in this when touching the existence rules. I think it's "rule-global", so you can't actually do this, or you need to have two rules.

Reply:
ping @ndmitchell

Reply:
More evidence can be obtained easily via shake profiling and ghcide-bench. The following experiment:
```
cabal build ghcide-bench
mkdir /tmp/ghcide
cabal exec cabal run ghcide-bench -- -- --example-package-name Cabal --example-module Distribution/Simple.hs --example-package-version 3.0.0.0 -s "edit" --samples 1 --shake-profiling /tmp/ghcide --no-clean
```
Then open the latest html file in `/tmp/ghcide` and check the number of `GetModificationTime` rules executed in the last run.

BEFORE: 471
AFTER: 238 (for the `.hi` files)

<img width="1170" alt="image" src="https://user-images.githubusercontent.com/26626/110205325-4b04af80-7e6f-11eb-8e20-335204acc62e.png">

Reply:
Thanks for checking, that looks convincing!